File: PLH-python/src/framework.py
from langchain_together import ChatTogether
from langchain.prompts import (
    ChatPromptTemplate,
    SystemMessagePromptTemplate,
    PromptTemplate,
)
from langchain_openai import ChatOpenAI
from src.conversation_tracer import ConversationTracer, TraceEntry
from src.config import Config
from src.decision_types import DecisionType
from dotenv import load_dotenv
import os
import yaml
import json
from typing import List, Optional, Tuple
from datetime import datetime

# Load environment variables from .env file
load_dotenv()


class Framework:
    def __init__(self, config, debug_mode=False):
        if not os.getenv("TOGETHER_API_KEY"):
            raise ValueError(
                "TOGETHER_API_KEY environment variable is not set. Please check your .env file."
            )

        self.config = config
        self.debug_mode = debug_mode
        self.child_llm = ChatTogether(
            model=config.get("models", "child"),
            temperature=config.get("models", "child_temperature"),
        )
        self.facilitator_llm = ChatTogether(
            model=config.get("models", "facilitator"),
            temperature=config.get("models", "facilitator_temperature"),
        )

        self.child_system_prompt = config.get("system_prompts", "child")
        self.facilitator_decision_system_prompt = config.get(
            "system_prompts", "facilitator_decision"
        )
        self.facilitator_positive_reinforcement_system_prompt = config.get(
            "system_prompts", "facilitator_positive_reinforcement"
        )
        self.facilitator_help_system_prompt = config.get(
            "system_prompts", "facilitator_help"
        )
        self.facilitator_end_coaching_system_prompt = config.get(
            "system_prompts", "facilitator_end_coaching"
        )
        self.facilitator_summary_system_prompt = config.get(
            "system_prompts", "facilitator_summary"
        )

        self.child_pipeline = (
            ChatPromptTemplate.from_messages(
                [SystemMessagePromptTemplate.from_template(self.child_system_prompt)]
            ).partial(
                scenario_description=self.config.get("scenario", "description"),
                # scenario_objectives=self.config.get("scenario", "objectives"),
            )
            | self.child_llm
        )

        self.facilitator_decision_pipeline = (
            ChatPromptTemplate.from_messages(
                [
                    SystemMessagePromptTemplate.from_template(
                        self.facilitator_decision_system_prompt
                    )
                ]
            ).partial(
                scenario_description=self.config.get("scenario", "description"),
                scenario_objectives=self.config.get("scenario", "objectives"),
                end_conversation=self.config.get("conditions", "end_conversation"),
                child_only_neutral=self.config.get("conditions", "child_only_neutral"),
                child_only_positive=self.config.get(
                    "conditions", "child_only_positive"
                ),
                child_and_facilitator_positive_reinforcement=self.config.get(
                    "conditions", "child_and_facilitator_positive_reinforcement"
                ),
                child_and_facilitator_help=self.config.get(
                    "conditions", "child_and_facilitator_help"
                ),
                facilitator_only_help=self.config.get(
                    "conditions", "facilitator_only_help"
                ),
            )
            | self.facilitator_llm
        )

        self.facilitator_positive_reinforcement_pipeline = (
            ChatPromptTemplate.from_messages(
                [
                    SystemMessagePromptTemplate.from_template(
                        self.facilitator_positive_reinforcement_system_prompt
                    )
                ]
            ).partial(
                scenario_description=self.config.get("scenario", "description"),
                scenario_objectives=self.config.get("scenario", "objectives"),
            )
            | self.facilitator_llm
        )

        self.facilitator_help_pipeline = (
            ChatPromptTemplate.from_messages(
                [
                    SystemMessagePromptTemplate.from_template(
                        self.facilitator_help_system_prompt
                    )
                ]
            ).partial(
                scenario_description=self.config.get("scenario", "description"),
                scenario_objectives=self.config.get("scenario", "objectives"),
            )
            | self.facilitator_llm
        )

        self.facilitator_end_coaching_pipeline = (
            ChatPromptTemplate.from_messages(
                [
                    SystemMessagePromptTemplate.from_template(
                        self.facilitator_end_coaching_system_prompt
                    )
                ]
            ).partial(
                scenario_description=self.config.get("scenario", "description"),
                scenario_objectives=self.config.get("scenario", "objectives"),
            )
            | self.facilitator_llm
        )

        self.facilitator_summary_pipeline = (
            ChatPromptTemplate.from_messages(
                [
                    SystemMessagePromptTemplate.from_template(
                        self.facilitator_summary_system_prompt
                    )
                ]
            ).partial(
                scenario_description=self.config.get("scenario", "description"),
                scenario_objectives=self.config.get("scenario", "objectives"),
            )
            | self.facilitator_llm
        )

        self.conversation_trace = ConversationTracer()
        self.turn_count = 1

    def _debug_print(self, prompt_name: str, prompt_content: str):
        """Helper method to print debug information if debug mode is enabled."""
        if self.debug_mode:
            print(f"\n=== {prompt_name} ===")
            print(prompt_content)
            print("===========================\n")

    def generate_child_response(self, parent_input):
        prompt_inputs = {
            "parent_response": parent_input,
            "interaction_history": self.conversation_trace.get_pretty_conversation(),
            "scenario_description": self.config.get("scenario", "description"),
            # "scenario_objectives": self.config.get("scenario", "objectives"),
            "turn_count": self.turn_count,
        }

        if self.debug_mode:
            constructed_prompt = self.child_system_prompt.format(**prompt_inputs)
            self._debug_print("Child Response Prompt", constructed_prompt)

        child_response = self.child_pipeline.invoke(prompt_inputs)
        return child_response.content

    def log_interaction(self, parent, child, decision, decision_reasoning, coaching):
        entry = TraceEntry(parent, child, decision, decision_reasoning, coaching)
        self.conversation_trace.add_entry(entry)

    def generate_decision(self, parent_input, child_response=None) -> tuple[int, str]:
        prompt_inputs = {
            "parent_response": parent_input,
            "child_response": self.conversation_trace.get_latest_child_message(),
            "interaction_history": self.conversation_trace.get_pretty_trace_full(
                exclude_latest_child=True
            ),
            "turn_count": self.turn_count,
            "scenario_description": self.config.get("scenario", "description"),
            "scenario_objectives": self.config.get("scenario", "objectives"),
            "end_conversation": self.config.get("conditions", "end_conversation"),
            "child_only_neutral": self.config.get("conditions", "child_only_neutral"),
            "child_only_positive": self.config.get("conditions", "child_only_positive"),
            "child_and_facilitator_positive_reinforcement": self.config.get(
                "conditions", "child_and_facilitator_positive_reinforcement"
            ),
            "child_and_facilitator_help": self.config.get(
                "conditions", "child_and_facilitator_help"
            ),
            "facilitator_only_help": self.config.get(
                "conditions", "facilitator_only_help"
            ),
        }

        if self.debug_mode:
            constructed_prompt = self.facilitator_decision_system_prompt.format(
                **prompt_inputs
            )
            self._debug_print("Facilitator Decision Prompt", constructed_prompt)

        max_retries = 3
        attempts = 0

        while attempts < max_retries:
            attempts += 1

            facilitator_response = self.facilitator_decision_pipeline.invoke(
                prompt_inputs
            )

            decision = None
            feedback = ""
            decision_found = False
            reasoning_found = False

            for line in facilitator_response.content.split("\n"):
                if line.startswith("DECISION:"):
                    decision_found = True
                    try:
                        decision = int(line.split(":")[1].strip())
                    except (ValueError, IndexError):
                        if self.debug_mode:
                            print(f"Invalid decision format: {line}")
                        continue
                elif line.startswith("REASONING:"):
                    reasoning_found = True
                    feedback = line.replace("REASONING:", "", 1).strip()

            try:
                if not decision_found or decision is None:
                    raise ValueError("Decision not found in the response")
                if not reasoning_found or not feedback:
                    raise ValueError("Reasoning not found in the response")

                DecisionType(decision)

                return (decision, feedback)

            except ValueError as e:
                if self.debug_mode:
                    print(f"Validation error (attempt {attempts}/{max_retries}): {e}")

                if attempts >= max_retries:
                    valid_values = [e.value for e in DecisionType]
                    raise ValueError(
                        f"Failed to get valid decision after {max_retries} attempts. "
                        f"Valid values are: {valid_values}. Last error: {str(e)}"
                    )

                continue

    def generate_positive_coaching(
        self,
        parent_input,
        child_response=None,
        reasoning=None,
        facilitator_only_response=False,
    ):
        prompt_inputs = {
            "parent_response": parent_input,
            "child_response": self.conversation_trace.get_latest_child_message(),
            # "child_response": child_response,
            "interaction_history": self.conversation_trace.get_pretty_trace_full(
                exclude_latest_child=True
            ),
            "scenario_description": self.config.get("scenario", "description"),
            "scenario_objectives": self.config.get("scenario", "objectives"),
            "previous_coaching": self.conversation_trace.get_previous_coaching(),
            "reasoning": reasoning,
        }

        if self.debug_mode:
            constructed_prompt = (
                self.facilitator_positive_reinforcement_system_prompt.format(
                    **prompt_inputs
                )
            )
            self._debug_print(
                "Facilitator Positive Reinforcement Prompt", constructed_prompt
            )

        facilitator_coaching_feedback = (
            self.facilitator_positive_reinforcement_pipeline.invoke(prompt_inputs)
        )
        if facilitator_only_response:
            return (
                self.config.get("static_messages", "retry_message")
                + " "
                + facilitator_coaching_feedback.content
            )
        return facilitator_coaching_feedback.content

    def generate_negative_coaching(
        self, parent_input, reasoning, facilitator_only_response=False
    ):
        prompt_inputs = {
            "parent_response": parent_input,
            "child_response": self.conversation_trace.get_latest_child_message(),
            # "child_response": child_response,
            "interaction_history": self.conversation_trace.get_pretty_trace_full(
                exclude_latest_child=True
            ),
            "scenario_description": self.config.get("scenario", "description"),
            "scenario_objectives": self.config.get("scenario", "objectives"),
            "reasoning": reasoning,
        }

        if self.debug_mode:
            constructed_prompt = self.facilitator_help_system_prompt.format(
                **prompt_inputs
            )
            self._debug_print("Facilitator Help Prompt", constructed_prompt)

        facilitator_coaching_feedback = self.facilitator_help_pipeline.invoke(
            prompt_inputs
        )
        if facilitator_only_response:
            return (
                self.config.get("static_messages", "retry_message")
                + " "
                + facilitator_coaching_feedback.content
            )
        return facilitator_coaching_feedback.content

    def generate_end_coaching(self, parent_input, reasoning):
        """Generate coaching feedback when the conversation is ending."""
        prompt_inputs = {
            "parent_response": parent_input,
            "child_response": self.conversation_trace.get_latest_child_message(),
            "interaction_history": self.conversation_trace.get_pretty_trace_full(
                exclude_latest_child=True
            ),
            "scenario_description": self.config.get("scenario", "description"),
            "scenario_objectives": self.config.get("scenario", "objectives"),
            "previous_coaching": self.conversation_trace.get_previous_coaching(),
            "reasoning": reasoning,
        }

        if self.debug_mode:
            constructed_prompt = self.facilitator_end_coaching_system_prompt.format(
                **prompt_inputs
            )
            self._debug_print("Facilitator End Coaching Prompt", constructed_prompt)

        # Use the pre-defined pipeline
        facilitator_coaching_feedback = self.facilitator_end_coaching_pipeline.invoke(
            prompt_inputs
        )
        return facilitator_coaching_feedback.content

    def generate_summary(self, parent_feedback_positive, parent_feedback_negative):
        prompt_inputs = {
            "interaction_history": self.conversation_trace.get_pretty_trace_full(),
            "scenario_description": self.config.get("scenario", "description"),
            "scenario_objectives": self.config.get("scenario", "objectives"),
            "parent_feedback_positive": parent_feedback_positive,
            "parent_feedback_negative": parent_feedback_negative,
        }

        if self.debug_mode:
            constructed_prompt = self.facilitator_summary_system_prompt.format(
                **prompt_inputs
            )
            self._debug_print("Facilitator Summary Prompt", constructed_prompt)

        facilitator_summary = self.facilitator_summary_pipeline.invoke(prompt_inputs)
        return facilitator_summary.content

File: PLH-node/src/framework.js
import ConfigHelper from "./configHelper.js";
import { ChatTogetherAI } from "@langchain/community/chat_models/togetherai";
import { SystemMessagePromptTemplate, ChatPromptTemplate } from "@langchain/core/prompts";

/**
 * Framework class handling the chat interactions between child and tutor models
 */
class Framework {
  constructor(lng) {
    const configHelper = new ConfigHelper(lng);
    this.config = configHelper.getConfig();
    this.#initializeModels();
    this.#initializePipelines();
  }

  /**
   * -----------------------------------------------------------------------------
   * MODEL INITIALIZATION
   * -----------------------------------------------------------------------------
   */

  #initializeModels() {
    this.#initializeChildModel();
    this.#initializeTutorModel();
  }

  #initializeChildModel() {
    this.childLLM = new ChatTogetherAI({
      model: this.config.models.child,
      temperature: this.config.models.child_temperature,
      apiKey: process.env.TOGETHER_AI_API_KEY,
    });
  }

  #initializeTutorModel() {
    this.tutorLLM = new ChatTogetherAI({
      model: this.config.models.facilitator,
      temperature: this.config.models.facilitator_temperature,
      apiKey: process.env.TOGETHER_AI_API_KEY,
    });
  }

  /**
   * -----------------------------------------------------------------------------
   * PIPELINE INITIALIZATION
   * -----------------------------------------------------------------------------
   */

  #initializePipelines() {
    this.#initializeChildPipeline();
    this.#initializeFacilitatorDecisionPipeline();
    this.#initializeFacilitatorPositiveReinforcementPipeline();
    this.#initializeHelpPipeline();
    this.#initializeFacilitatorEndCoachingPipeline();
    this.#initializeTutorSummaryPipeline();
  }

  #initializeChildPipeline() {
    this.childSystemMessage = SystemMessagePromptTemplate.fromTemplate(
      this.config.system_prompts.child,
      ["scenario_description"]
    );

    this.childPipeline = ChatPromptTemplate.fromMessages([this.childSystemMessage]).pipe(
      this.childLLM
    );
  }

  #initializeFacilitatorDecisionPipeline() {
    this.tutorDecisionMessage = SystemMessagePromptTemplate.fromTemplate(
      this.config.system_prompts.facilitator_decision,
      [
        "scenario_description",
        "scenario_objectives",
        "end_conversation",
        "child_only_neutral",
        "child_only_positive",
        "child_and_facilitator_positive_reinforcement",
        "child_and_facilitator_help",
        "facilitator_only_help",
      ]
    );

    this.tutorDecisionPipeline = ChatPromptTemplate.fromMessages([this.tutorDecisionMessage]).pipe(
      this.tutorLLM
    );
  }

  #initializeFacilitatorPositiveReinforcementPipeline() {
    this.tutorCoachMessage = SystemMessagePromptTemplate.fromTemplate(
      this.config.system_prompts.facilitator_positive_reinforcement,
      ["scenario_description", "scenario_objectives"]
    );

    this.tutorCoachPipeline = ChatPromptTemplate.fromMessages([this.tutorCoachMessage]).pipe(
      this.tutorLLM
    );
  }

  #initializeHelpPipeline() {
    this.tutorCoachMessage = SystemMessagePromptTemplate.fromTemplate(
      this.config.system_prompts.facilitator_help,
      ["scenario_description", "scenario_objectives"]
    );

    this.facilitatorHelpPipeline = ChatPromptTemplate.fromMessages([this.tutorCoachMessage]).pipe(
      this.tutorLLM
    );
  }

  #initializeFacilitatorEndCoachingPipeline() {
    this.tutorCoachMessage = SystemMessagePromptTemplate.fromTemplate(
      this.config.system_prompts.facilitator_end_coaching,
      ["scenario_description", "scenario_objectives"]
    );

    this.facilitatorEndCoachingPipeline = ChatPromptTemplate.fromMessages([
      this.tutorCoachMessage,
    ]).pipe(this.tutorLLM);
  }

  #initializeTutorSummaryPipeline() {
    this.tutorSummaryMessage = SystemMessagePromptTemplate.fromTemplate(
      this.config.system_prompts.facilitator_summary,
      ["scenario_description", "scenario_objectives", "interaction_history"]
    );

    this.tutorSummaryPipeline = ChatPromptTemplate.fromMessages([this.tutorSummaryMessage]).pipe(
      this.tutorLLM
    );
  }

  /**
   * -----------------------------------------------------------------------------
   * CHILD MODEL INTERACTION
   * -----------------------------------------------------------------------------
   */

  /**
   * Prepare variables for child response prompt
   * @param {string} parentInput - Parent's input message
   * @param {string} interactionHistory - Previous conversation history
   * @returns {Object} Variables object for prompt template
   */
  #prepareChildVariables(parentInput, interactionHistory, turnCount) {
    return {
      parent_response: parentInput,
      scenario_description: this.config.scenario.description,
      interaction_history: interactionHistory,
      turn_count: turnCount,
      //scenario_objectives: scenario.objectives,
    };
  }

  /**
   * Get response from child model based on parent input and interaction history
   * @param {string} parentInput - Parent's input message
   * @param {string} interactionHistory - Previous conversation history
   * @returns {string} Child's response
   */
  async getChildResponse(parentInput, interactionHistory, turnCount) {
    const variables = this.#prepareChildVariables(parentInput, interactionHistory, turnCount);

    // Debug: Format the prompt to see what will be sent to the LLM
    // const formattedMessage = await this.childSystemMessage.format(variables);
    // console.log("Formatted child prompt content:", formattedMessage.content);
    console.log("----------------getChildResponse:---------------- :", variables);
    const response = await this.childPipeline.invoke(variables);
    return response.content;
  }

  /**
   * -----------------------------------------------------------------------------
   * TUTOR MODEL INTERACTIONS
   * -----------------------------------------------------------------------------
   */

  /**
   * Prepare variables for tutor decision prompt
   * @param {string} parentInput - Parent's input message
   * @param {string} childResponse - Child's response message
   * @param {string} interactionHistory - Previous conversation history
   * @param {number} turnCount - Current turn count
   * @returns {Object} Variables object for prompt template
   */
  #prepareTutorDecisionVariables(
    parentInput,
    childResponse,
    interactionHistory = "",
    turnCount = 1
  ) {
    return {
      parent_response: parentInput,
      child_response: childResponse,
      interaction_history: interactionHistory,
      turn_count: turnCount,
      scenario_description: this.config.scenario.description,
      scenario_objectives: this.config.scenario.objectives,
      end_conversation: this.config.conditions.end_conversation,
      child_only_neutral: this.config.conditions.child_only_neutral,
      child_only_positive: this.config.conditions.child_only_positive,
      child_and_facilitator_positive_reinforcement:
        this.config.conditions.child_and_facilitator_positive_reinforcement,
      child_and_facilitator_help: this.config.conditions.child_and_facilitator_help,
      facilitator_only_help: this.config.conditions.facilitator_only_help,
    };
  }

  /**
   * Get decision from tutor model based on parent and child interaction
   * @param {string} parentInput - Parent's input message
   * @param {string} childResponse - Child's response message
   * @param {string} interactionHistory - Previous conversation history
   * @param {number} turnCount - Current turn count
   * @returns {Object} Decision object with decision code and reasoning
   */
  async getDecision(parentInput, childResponse, interactionHistory, turnCount) {
    const variables = this.#prepareTutorDecisionVariables(
      parentInput,
      childResponse,
      interactionHistory,
      turnCount
    );

    // Debug: Format the prompt to see what will be sent to the LLM
    // const formattedMessage = await this.tutorDecisionMessage.format(variables);
    // console.log("Formatted decision prompt:", formattedMessage.content);
    console.log("----------------getDecision:---------------- :", variables);
    const response = await this.tutorDecisionPipeline.invoke(variables);

    return this.#parseDecisionResponse(response.content);
  }

  /**
   * Parse the decision response from the tutor model
   * @param {string} responseContent - Raw response content from model
   * @returns {Object} Parsed decision with decision code and reasoning
   * @throws {Error} If decision code is invalid
   */
  #parseDecisionResponse(responseContent) {
    let decision = null;
    let reasoning = "";

    // Extract decision and reasoning from response
    const lines = responseContent.split("\n");
    for (const line of lines) {
      if (line.startsWith("DECISION:")) {
        decision = parseInt(line.split(":")[1].trim());
      } else if (line.startsWith("REASONING:")) {
        reasoning = line.split(":")[1].trim();
      }
    }

    if (![0, 1, 2, 3, 4, 5].includes(decision)) {
      throw new Error(`Invalid decision value: ${decision}`);
    }

    return {
      decision,
      reasoning,
      raw: responseContent,
    };
  }

  /**
   * Prepare variables for tutor coaching prompt
   * @param {string} parentInput - Parent's input message
   * @param {string} childResponse - Child's response message
   * @param {string} interactionHistory - Previous conversation history
   * @returns {Object} Variables object for prompt template
   */
  #preparePositiveCoachVariables(
    parentInput,
    childResponse,
    interactionHistory = "",
    previousCoaching,
    reasoning
  ) {
    return {
      parent_response: parentInput,
      child_response: childResponse,
      interaction_history: interactionHistory,
      scenario_description: this.config.scenario.description,
      scenario_objectives: this.config.scenario.objectives,
      previous_coaching: previousCoaching,
      reasoning: reasoning,
    };
  }

  /**
   * Get coaching feedback from tutor model
   * @param {string} parentInput - Parent's input message
   * @param {string} childResponse - Child's response message
   * @param {string} interactionHistory - Previous conversation history
   * @returns {string} Coaching feedback
   */
  async getPositiveCoachingFeedback(
    parentInput,
    childResponse,
    interactionHistory,
    previousCoaching,
    reasoning,
    tutorOnlyResponse = false
  ) {
    if (!this.tutorCoachPipeline) {
      return "";
    }

    const variables = this.#preparePositiveCoachVariables(
      parentInput,
      childResponse,
      interactionHistory,
      previousCoaching,
      reasoning
    );
    console.log("----------------getPositiveCoachingFeedback:---------------- :", variables);
    const response = await this.tutorCoachPipeline.invoke(variables);
    if (tutorOnlyResponse) {
      return response.content + ` ${this.config.static_messages.retry_message}`;
    }
    return response.content;
  }

  /**
   * Prepare variables for tutor coaching prompt
   * @param {string} parentInput - Parent's input message
   * @param {string} childResponse - Child's response message
   * @param {string} interactionHistory - Previous conversation history
   * @param {string} reasoning - Previous reasoning history
   * @returns {Object} Variables object for prompt template
   */
  #prepareNegativeCoachVariables(parentInput, childResponse, interactionHistory = "", reasoning) {
    return {
      parent_response: parentInput,
      child_response: childResponse,
      interaction_history: interactionHistory,
      scenario_description: this.config.scenario.description,
      scenario_objectives: this.config.scenario.objectives,
      reasoning: reasoning,
    };
  }

  /**
   * Get coaching feedback from tutor model
   * @param {string} parentInput - Parent's input message
   * @param {string} childResponse - Child's response message
   * @param {string} interactionHistory - Previous conversation history
   * @returns {string} Coaching feedback
   */
  async getNegativeCoachingFeedback(
    parentInput,
    childResponse,
    interactionHistory,
    reasoning,
    tutorOnlyResponse = false
  ) {
    if (!this.facilitatorHelpPipeline) {
      return "";
    }

    const variables = this.#prepareNegativeCoachVariables(
      parentInput,
      childResponse,
      interactionHistory,
      reasoning
    );
    console.log("----------------getNegativeCoachingFeedback:---------------- :", variables);
    const response = await this.facilitatorHelpPipeline.invoke(variables);
    if (tutorOnlyResponse) {
      return response.content + ` ${this.config.static_messages.retry_message}`;
    }
    return response.content;
  }

  /**
   * Prepare variables for tutor coaching prompt
   * @param {string} parentInput - Parent's input message
   * @param {string} childResponse - Child's response message
   * @param {string} interactionHistory - Previous conversation history
   * @param {string} reasoning - Previous reasoning history
   * @returns {Object} Variables object for prompt template
   */
  #prepareEndCoachVariables(
    parentInput,
    childResponse,
    interactionHistory = "",
    previousCoaching,
    reasoning
  ) {
    return {
      parent_response: parentInput,
      child_response: childResponse,
      interaction_history: interactionHistory,
      scenario_description: this.config.scenario.description,
      scenario_objectives: this.config.scenario.objectives,
      previous_coaching: previousCoaching,
      reasoning: reasoning,
    };
  }

  /**
   * Get coaching feedback from tutor model
   * @param {string} parentInput - Parent's input message
   * @param {string} childResponse - Child's response message
   * @param {string} interactionHistory - Previous conversation history
   * @returns {string} Coaching feedback
   */
  async getEndCoachingFeedback(
    parentInput,
    childResponse,
    interactionHistory,
    previousCoaching,
    reasoning
  ) {
    if (!this.facilitatorHelpPipeline) {
      return "";
    }

    const variables = this.#prepareEndCoachVariables(
      parentInput,
      childResponse,
      interactionHistory,
      previousCoaching,
      reasoning
    );
    console.log("----------------getEndCoachingFeedback----------------: ", variables);
    const response = await this.facilitatorEndCoachingPipeline.invoke(variables);
    return response.content;
  }

  /**
   * Get conversation summary from tutor model
   * @param {string} interactionHistory - Complete conversation history
   * @returns {string} Conversation summary
   */
  async getConversationSummary(interactionHistory, positiveFeedback, negativeFeedback) {
    if (!this.tutorSummaryPipeline) {
      return "";
    }

    const variables = {
      interaction_history: interactionHistory,
      scenario_description: this.config.scenario.description,
      scenario_objectives: this.config.scenario.objectives,
      parent_feedback_positive: positiveFeedback,
      parent_feedback_negative: negativeFeedback,
    };
    console.log("----------------getConversationSummary:---------------- :", variables);
    const response = await this.tutorSummaryPipeline.invoke(variables);
    return response.content;
  }
}

export default Framework;
